### ISSUE TITLE
Add support for MetadataTypeAttribute

### ISSUE THREAD

**Issue by daniel-white** [destructurama/attributed#40]
[`System.ComponentModel.DataAnnotations.MetadataTypeAttribute` ](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.dataannotations.metadatatypeattribute?view=netcore-3.1) is a handy bridge to attributes defined in other files, especially in code generated files. It works with the validation frameworks. I don't see why it wouldn't here.

Example:
```
[GeneratedCode]
[MetadataTypeAttribute(typeof(DtoMetadata))]
public partial class Dto
{
    public string Private { get; set; }
}

internal class DtoMetadata
{
     [NotLogged]
     public object Private { get; set; }
}
```

I think it would be great if both sources could be utilized, but the configuration on the primary class would win in the event of a conflict.

**Comment by sungam3r** [comment]
What problem do you solve?

**Comment by daniel-white** [comment]
> What problem do you solve?

It solves one case where the properties are defined in a code generated class file. One wouldn’t want to put the attributes in there as they’d get overwritten by the generator

**Comment by sungam3r** [comment]
ok

**Comment by daniel-white** [comment]
@nblumhardt  is this something you would take a PR for? I'd be happy to start looking how to add this.

I've found this as a potential option: https://stackoverflow.com/questions/17873576/metadata-were-not-loaded-using-metadatatype

**Comment by nblumhardt** [comment]
Hi @daniel-white  - I think so :+1:

**Comment by daniel-white** [comment]
doh. the attribute isn't available in .net standard, so i could add one to the library and reflect on that or just wait till everything is .net 5.

**Comment by nblumhardt** [comment]
Targeting netcoreapp3.0 or 3.1 would also be fine 👍

**Comment by digitalsigi** [comment]
I am scaffolding repeatedly a database structure and separated all attributes in a separate file. So I have exactly the same problem.

@nblumhardt May be we can somehow work together to get it done. I am not experienced in multi-targeting, so I would need some help here.

According to [Micosoft](https://docs.microsoft.com/en-us/dotnet/standard/net-standard#when-to-target-net50-vs-netstandard) it sounds to go straight to net5.0.

**Comment by SimonCropp** [comment]
@digitalsigi are you still interested in submitting a PR for this?

**Comment by digitalsigi** [comment]
Unfortunately not for the moment. I am still busy with a larger project. Had a quick look into the code, it looks feasable, but its more than just hack an go. If it helps, I could provide some code snippets showing how to dig into attributes hierarchies.

**Comment by digitalsigi** [comment]
OK, got a ping from github. Here a code snippet I am using to get all properties which are then filtered to select only those with my Custom Attribute:
```c#
           var propBaseType = vmType.GetProperties().ToList();
           var metaType = vmType.GetCustomAttributes<MetadataTypeAttribute>(true).ToList().FirstOrDefault();
           var metaProp = new List<PropertyInfo>();
           if (metaType != null)
           {
               var metaClass = metaType.MetadataClassType;
               metaProp = metaClass.GetProperties().ToList();
           }
 
           var propConcat = metaProp.Concat(propBaseType) // Concatenate properties of MetaData class and Base class
                                                          // Take only first -> the one of Metadata, assumed metadata has the avs attribute
                                                          // if base class bears an avs attibute it is silently ignored
                                   .OrderBy(m => m.Name)
                                   ;

```

I guess something like this has to be added in GetPropertiesRecursive, but I still can't find the time to do it.

**Comment by sungam3r** [comment]
It definetely can be added. The main thing here is to match attributes from other type to the properties of original type.

**Comment by sungam3r** [comment]
If someone wants to work on this issue, I will be grateful.

**Comment by sungam3r** [comment]
I don't think I will work on that features. Anyone can start up from #91.

**Comment by sungam3r** [comment]
ping @daniel-white @digitalsigi

**Comment by digitalsigi** [comment]
Ok folks, 

I just synced my fork and I'll have at look at it.

@sungam3r It looks like you are working on something to provide a solution. Whats your status?

**Comment by sungam3r** [comment]
Nope. See https://github.com/destructurama/attributed/issues/40#issuecomment-1913513640

**Comment by digitalsigi** [comment]
Just saw a pending PR. I will hab a look but possibly open a new PR is this OK for you?

**Comment by sungam3r** [comment]
Yes.
